src/robot/libdocpkg/model.py

```diff
@@ -190,3 +228,155 @@ def _arg_to_dict(self, arg):
+        self.doc = doc
+        self.items = items or []
+
+    @classmethod
+    def from_dict(cls, type_doc):
+        if isinstance(type_doc, dict):
+            return cls(name=type_doc['name'],
+                       type=type_doc['type'],
+                       doc=type_doc['doc'],
+                       items=type_doc['items'])
+        raise TypeError(
+            'TypedDictDoc.from_dict() requires dictionary types but got %s.'
+            % type_name(type_doc))
+
+    @classmethod
+    def from_TypedDict(cls, typed_dict):
+        if isinstance(typed_dict, (TypedDictType, ExtTypedDictType)):
+            items = []
+            required_keys = list(getattr(typed_dict, '__required_keys__', []))
+            optional_keys = list(getattr(typed_dict, '__optional_keys__', []))
+            for key, value in typed_dict.__annotations__.items():
+                typ = value.__name__ if isclass(value) else unic(value)
+                items.append({'key': key,
+                              'type': typ,
+                              'required': key in required_keys
+                              if required_keys or optional_keys else None})
+            return cls(name=typed_dict.__name__,
+                       type='TypedDict',
+                       doc=typed_dict.__doc__ if typed_dict.__doc__ else '',
+                       items=items)
+        raise TypeError(
+            'TypedDictDoc.from_TypedDict() requires a TypedDict but got %s.'
+            % type_name(typed_dict))
+
+    @property
+    def _sort_key(self):
+        return self.name.lower()
+
+    def to_dictionary(self):
+        return {
+            'name': self.name,
+            'type': self.type,
+            'doc': self.doc,
+            'items': self.items
+        }
+
+
+class EnumDoc(Sortable):
+
+    def __init__(self, name='', type='', doc='', members=None):
+        self.name = name
+        self.type = type
+        self.doc = doc
+        self.members = members or []
+
+    @classmethod
+    def from_dict(cls, type_doc):
+        if isinstance(type_doc, dict):
+            return cls(name=type_doc['name'],
+                       type=type_doc['type'],
+                       doc=type_doc['doc'],
+                       members=type_doc['members'])
+        raise TypeError(
+            'EnumDoc.from_dict() requires dictionary types but got %s.'
+            % type_name(type_doc))
+
+    @classmethod
+    def from_Enum(cls, enum_type):
+        if isinstance(enum_type, EnumType):
+            return cls(name=enum_type.__name__,
+                       type='Enum',
+                       doc=enum_type.__doc__ or '',
+                       members=[{'name': name,
+                                 'value': unicode(member.value)}
+                                for name, member in enum_type.__members__.items()])
+        raise TypeError(
+            'EnumDoc.from_Enum() requires Enum types but got %s.'
+            % type_name(enum_type))
```

Comment:
Are you sure this kind of error handling is needed? How likely is it that someone calls `from_TypedDict` or `from_dict` with an invalid argument? If such error handling really is needed, it should be done like

```python
if not valid:
    raise TypeError(...)
# several lines of code
```

not like
```python
if valid:
    # several lines of code
    return
raise TypeError(...)
```